File: 1100-find-k-length-substrings-with-no-repeated-characters/1100-find-k-length-substrings-with-no-repeated-characters.cpp
class Solution {
public:
    int numKLenSubstrNoRepeats(string s, int k) {
        int start = 0, end = 0; 
        int answer = 0;  
        unordered_map<char,int> hashMap; 
        while(end < s.length()){
            hashMap[s[end]]++; 
            while(hashMap[s[end]] > 1){
                hashMap[s[start]]--;
                start++; 
            }

            if(end - start + 1 == k){
                answer++; 
                hashMap[s[start]]--;
                start++; 
            }
            end++; 
        }
        return answer; 
    }
};

File: 1100-find-k-length-substrings-with-no-repeated-characters/1100-find-k-length-substrings-with-no-repeated-characters.java
class Solution {
    public int numKLenSubstrNoRepeats(String s, int k) {
        Map<Character,Integer> hashMap = new HashMap<>(); 
        int start = 0, end = 0, answer = 0;  
        while(end < s.length()){
            char curr = s.charAt(end); 
            while(hashMap.containsKey(curr) || end - start + 1 > k){
                char prev = s.charAt(start); 
                hashMap.merge(prev, -1, Integer::sum); 
                if(hashMap.get(prev) <= 0) hashMap.remove(prev); 
                start++; 
            }

            if(end - start + 1 == k){
                answer++; 
            }

            hashMap.merge(curr,1,Integer::sum); 
            end++; 
        }

        return answer; 
    }
}